### ISSUE TITLE
Implement fetching of module ID

### ISSUE THREAD

**Issue by kipe** [kipe/enocean#15]
Every message should be send with the correct module ID, from which the message originates.
A simple test implementing this is created in f6825c45161600ac3dec95c37a9e0b7a94ca300a. However, it's not tested and thus isn't included in the `Communicator` initialization.

Once tested, the module ID should be fetched when initializing the `Communicator`. If ID is not set, sending of messages should be prevented.

I currently have no hardware for testing this, so help wanted.

**Comment by romor** [comment]
I am no expert in EnOcean protocol but want to share my observations with regard to addressing:

Devices used:
 - multiple Micropelt iTRV HVAC actuators
 - one single Thermokon STC-MSG heating controller
 - Thermokon SR06 temperature sensors (not relevant)

What I recognize:
When the STC-MSG send a message to the iTRV it sets a different source address for its messages, depending on the target iTRV to address (e.g. 0xFFD05081 for the first iTRV actuator and 0xFFD05082 for the second). And it sets the destination address in the packets to 0xffffffff, i.e. as broadcast. At least this is what I receive when I sniff the packets using this enocean library.

So, the different iTRV actuator devices are "addressed" using the source address of the packet in a broadcast packet. From this I conclude that limiting the source addresses to the module ID is not feasible, or at least not done in the setup above.

**Comment by kipe** [comment]
I think you've misunderstood the issue.
This is regarding the [`Packet.create` `sender`-input](https://github.com/kipe/enocean/blob/f6825c45161600ac3dec95c37a9e0b7a94ca300a/examples/enocean_example.py#L18). It's requested on [line 29](https://github.com/kipe/enocean/blob/f6825c45161600ac3dec95c37a9e0b7a94ca300a/examples/enocean_example.py#L29) from the module attached to your Raspberry Pi or the USB-dongle. (For documentation, see [EnOcean Serial Protocol 3](http://www.enocean.com/esp), page 34).

So in essence, the idea is not to limit the `destination` of `Packet.create`, just using the correct device ID, from which we're sending. So for example, you would set the `destination` to 0xFFD05081 or 0xFFD05082, and `sender` would be the ID of your transmitter/receiver module attached to your RasPi/PC.

**Comment by kipe** [comment]
Also, the situation is a bit different when sending or receiving packets. When receiving, the destination is set to broadcast and source is the device which sends the data.

The sending is done in a similar fashion, but of course we can set the destination to either target a specific device or broadcast to all devices. What I'd like to do, is to limit the setting of sender to the actual, unique ID of the module sending the message.

**Comment by romor** [comment]
No sure, I think I understood the idea of this ticket and it also sounds logical to me. However:

What I recognized by sniffing my EnOcean packets from the single STC-MSG device to 2 different iTRVs devices is: 
The two iTRVs send individual requests to the STC-MSG using their different source addresses (direction 1 of EEP A5-20-01) (of course, not synchronous and not simultanous). When the single STC-MSG device answers each request (direction 2 of A5-20-01), this packet had different **source** addresses of the single STR-MSG device, depending to which iTRV it was sent. 

Whether these packets have also different destination addresses which were masked then to broadcasts from the USB dongle (because this was shown in the packet dump as destination) - I do not know.
I assumed they were really sent to broadcast and the receiving iTRVs recognized whether they are affected by the (learnt-in) source address of the message.

Actually the single STC-MSG heating controller device uses 9 different **source** addresses: 1 address for each HVAC actuator it controls and and 9th one for sending its status (EEP A5-11-02).

**Comment by romor** [comment]
Just for intuition, the source addresses in my device are:
0xFFD05081 to 0xFFD05088 for the 8 channels and 0xFFD050E4 for the status message.

**Comment by kipe** [comment]
Hmmm, can you post a dump of the data? Because that kind of behavior seems very weird and illogical :S

**Comment by romor** [comment]
Unfortunately not. It was a demonstration device which I sent back again. The heating control is now done using this lib on a RPi...

Anyway, to conclude: From my perspective this ticket absolutely makes sense. I would just suggest to allow overriding the sender address, for the potential case that this is needed.

**Comment by kipe** [comment]
Well, can you then post the working code and (perhaps most importantly) the `Packet.build()` -results of controlling the heating?
This still bugs me, as it seems like a very, very strange implementation :confused:

**Comment by romor** [comment]
Sure, I can provide the logs of my current implementation:

Controlling HVAC1:

    2016-01-29 20:54:45,912 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-85 dBm): 0x01 ['0xa5', '0x0', '0x20', '0x85', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x55', '0x0'] {}
    2016-01-29 20:54:46,225 INFO: sending: 0x01 ['0xa5', '0x0', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x4c', '0x0'] ['0x3', '0xff', '0xff', '0xff', '0xff', '0xff', '0x0'] {}

Controlling HVAC2:

    2016-01-29 20:57:50,295 INFO: received: FF:DC:95:00->FF:FF:FF:FF (-74 dBm): 0x01 ['0xa5', '0x0', '0x20', '0x83', '0x8', '0xff', '0xdc', '0x95', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x4a', '0x0'] {}
    2016-01-29 20:57:50,586 INFO: sending: 0x01 ['0xa5', '0x0', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x70', '0x0'] ['0x3', '0xff', '0xff', '0xff', '0xff', '0xff', '0x0'] {}

Is this what you mean?
Actually it just uses the byte sum of the HVAC address as the last byte of the sender address (0x4c or 0x70) and addresses the HVAC actuators using broadcast addressing. It works fine.

But to be honest, I never tried to address the actuators individually with their target address and the same sender address because I just copied what I observed from the STC-MSG controller.

**Comment by kipe** [comment]
Could you try sending the following packets, for controlling HVAC1:
```python
p = Packet(0x01, data=[0xa5, 0x0, 0x79, 0x0, 0x8, 0x17, 0x11, 0x19, 0x4c, 0x0], optional=[0x3, 0xff, 0xff, 0xff, 0xff, 0xff, 0x0])
built_packet = p.build()
print(built_packet)
#  [85, 0, 10, 7, 1, 235, 165, 0, 121, 0, 8, 23, 17, 25, 76, 0, 3, 255, 255, 255, 255, 255, 0, 159]
```

```python
p = Packet(0x01, data=[0xa5, 0x0, 0x79, 0x0, 0x8, 0xDE, 0xAD, 0xBE, 0xEF, 0x0], optional=[0x3, 0xFF, 0xD5, 0x78, 0x00, 0xff, 0x0])
built_packet = p.build()
print(built_packet)
# [85, 0, 10, 7, 1, 235, 165, 0, 121, 0, 8, 222, 173, 190, 239, 0, 3, 255, 213, 120, 0, 255, 0, 191]
```

The first one should be the same as you were sending now (just for the purposes of checking). The second uses the device id the value is received from as the destination for the package created. The sender id is our beloved 0xDEADBEEF ;)

This is how I imagine the protocol should work, if it doesn't, I'm not sure what to do...
One idea that comes to mind is the Addressing Destination Telegram  (RORG.ADT), but I have no idea on how to actually use it.

**Comment by romor** [comment]
I guess it is not that simple to try because the HVAC actuator also needs to be learnt to the correct sender address. So I need to adapt the learning sequence also.

Anyway, I am already about trying this out. I expected your interest and you also raised mine...

**Comment by kipe** [comment]
Great :+1: 
I'm actually thinking of contacting EnOcean technical support related to this, if we can't figure this out. I still can't understand why have unique addresses, if they're not properly used :stuck_out_tongue:

**Comment by romor** [comment]
What I observed is:
 - the HVAC actuator reacts on a command independent on the packet's destination address. It reacts on broadcast, the correct address but also a random destination address.
 - the HVAC actuator also reacts on a command independent on the packet's sender address.

See the following log:

    2016-01-30 01:29:14,556 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-88 dBm): 0x01 ['0xa5', '0xf', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x58', '0x0'] {}
    2016-01-30 01:29:14,899 INFO: sending: 0x01 ['0xa5', '0x3', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x77', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}
    2016-01-30 01:29:17,483 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-86 dBm): 0x01 ['0xa5', '0xf', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x56', '0x0'] {}
    2016-01-30 01:29:17,724 INFO: sending: 0x01 ['0xa5', '0x6', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x77', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}

    2016-01-30 01:29:55,343 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-88 dBm): 0x01 ['0xa5', '0x6', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x58', '0x0'] {}
    2016-01-30 01:29:55,571 INFO: sending: 0x01 ['0xa5', '0x9', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x76', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}
    2016-01-30 01:29:58,271 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-86 dBm): 0x01 ['0xa5', '0x6', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x56', '0x0'] {}
    2016-01-30 01:29:58,500 INFO: sending: 0x01 ['0xa5', '0xc', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x77', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}

    2016-01-30 01:31:12,444 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-86 dBm): 0x01 ['0xa5', '0xc', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x56', '0x0'] {}
    2016-01-30 01:31:12,729 INFO: sending: 0x01 ['0xa5', '0xf', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x77', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}
    2016-01-30 01:31:15,366 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-86 dBm): 0x01 ['0xa5', '0xc', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x56', '0x0'] {}
    2016-01-30 01:31:15,601 INFO: sending: 0x01 ['0xa5', '0x12', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x76', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}

    2016-01-30 01:32:21,315 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-86 dBm): 0x01 ['0xa5', '0x12', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x56', '0x0'] {}
    2016-01-30 01:32:21,553 INFO: sending: 0x01 ['0xa5', '0x18', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x77', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}
    2016-01-30 01:32:24,242 INFO: received: FF:D5:78:00->FF:FF:FF:FF (-86 dBm): 0x01 ['0xa5', '0x12', '0x20', '0x7b', '0x8', '0xff', '0xd5', '0x78', '0x0', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x56', '0x0'] {}
    2016-01-30 01:32:24,482 INFO: sending: 0x01 ['0xa5', '0x1b', '0x79', '0x0', '0x8', '0x17', '0x11', '0x19', '0x76', '0x0'] ['0x3', '0x5', '0x6', '0x7', '0x8', '0xff', '0x0'] {}

Note:
 - the HVAC actuator is learnt to sender address 0x17111976. This correct address is used at every 3rd packet only, otherwise the wrong sender address ending with 0x77 is used.
 - the learning is done using 4BS teach-in in Variation 3 (according EnOcean EEP document, appendix). Thus, both sides should also store the Sender-ID.
 - the HVAC actuator requests its new position and expects this to be given in the response packet. In the request packet it also sends its current position (CV) as the first data byte, which allows to determine whether the command in the answer was accepted or not.
 - the target address of the response packet 0x05060708 in the test does not match the HVAC address 0xffd57800.
 - the HVAC actuator has an automatic polling each 10 minutes but also a polling on request (via a button on the actuator). This button was used for the tests. If the button is used, it sends the request twice, in automatic polling only once (I don't know why).
 - it takes a few seconds for the actuator to set the new desired position, because of this the CV is shown delayed.

What you see is:
 - the HVAC actuator follows the given target position in the first byte (you can also hear the motor moving) for each packet provided, independent on addressing. The packet just needs to come within one second after the request.

My conclusions:
This Micropelt HVAC actuator is not working as it should. As far as I know, it is a very new product and my devices are test samples form the manufacturer. I will also raise this question to Micropelt.

**Comment by kipe** [comment]
Great work, as usual :)
I think this might require the implementation of sending RORG.ADT (Addressed Data Telegram), which (as I understand) is basically the same message, with the data prefixed with RORG.ADT (thus, making the data 1 byte longer). The ADT should notify the receiver that the data we're actually sending contains the destination address.
Not sure on the implementation though, so before going through the trouble of testing various methods, I think it's better to ask for help from Micropelt / EnOcean.

**Comment by romor** [comment]
I did some more review of the EnOcean EEP document and concluded that my initial assumption seems to be correct: The HVAC receiver identifies the message based on the **sender's** address, which is learnt and stored during teach-in.

Please refer to Section 1.7 of the EEP document (Teach-in procedures):
> For this purpose of a determined relationship between transmitter and receiver each transmitting device has a unique Sender-ID which is part of each radio telegram.

Also, looking in Section 1.7.1 of EnOceanSerialProtocol3_V1.17.pdf, Table 4 states that the destination address is always 0xffffffff for packets except ADT, i.e. 4BS (RORG 0xA5), which is obviously used in my devices.

Therefore, if one "server device" wants to control multiple "clients" it needs to use different Sender-IDs to ensure that the clients do not react on a packet, which is intended for a different client.

I also raised this question to Micropelt and will post their response here once (and if) I get one.

**Comment by kipe** [comment]
Thanks for this, good job as always.
Makes me just wonder how common the support for ADT is, if there's even a reason to implement it at this point... I think I'll try and implement it anyways once I receive the plugs to test on, and see if they react to it correctly.

Do you think we should implement some kind of algorithm to generate a `Sender-ID`, based on the real device address, how the stars align, whether or not my moms dog is sick etc? Another option is to leave this to the user, making implementation of automatic teach-in impossible.

The generation of the `Sender-ID` should be random enough not to yield the same ID "with a reasonable" accuracy though.

**Comment by romor** [comment]
I think the default `Sender-ID` should be auto-detected, just the way as you already implemented it in the example.

Optionally there should be the possibility to provide a callback to `Communicator` to derive a different `Sender-ID` as a function of the default `Sender-ID` and the destination address of the packet to send: 
`def derive_sender_id(default_sender_id, destination_address)`.

**Comment by romor** [comment]
This is the answer I received from Micropelt, which was very informative to me:
 - the Micropelt device supports only RORG 0xA5 and no ADT
 - in EnOcean, the sender address can be either a Unique ID (manufacturer chip-ID) or an address based on the Base-ID
 - the Base-ID can be set and read by COMMAN_COMMAND CO_WR_IDBASE (07) and CO_RD_IDBASE (08) and must be in range 0xFF800000-0xFFFFFF80 with the last 7 bits set to zero
 - these seven bits are used to emulate 128 IDs and are usually incremented for each device learnt, so one module can emulate 128 senders using the Base-ID
 - the actuator learns this unique Sender-ID during teach-in and uses this for addressing

The problem in my test described above was, that I used the sender address 0x17111976, which is outside the valid Base-ID range.

So, I see two possible approaches for this enocean library:
 1. the callback specified above, which leaves the user the obligation to determine the address (or the offset to the base-ID) based on the destination address
 2. the enocean library maintains this mapping, which implies the need to store, read, reset,... this

**Comment by kipe** [comment]
Thanks to both @romor and Micropelt :+1: 

I was wondering the use of `Base-ID` before myself, as the number of writes seemed to be quite small. This clears up the question of writing it; there really shouldn't be any need.

**As for the possible approaches:**

_Number 1_
- Requires a lot less work.
- Leaves a lot to the end-user.
- Timings are quite crucial in EnOcean, for example the response to `UTE`-teach-in should be sent in 700ms (not sure if devices actually obey this).
- Because of timings, using databases on low-end devices like the Raspberry Pi (for which I originally started developing this for) might prove difficult (don't know, haven't tested database operation performances).
   - This affects me personally, as it's the setup I'm running on.

_Number 2_
- A lot more work for us.
  - At the current state of this library, we have much to implement as is. Not sure this is the first place we should focus on.
  - The structure of the storage requires planning. Do we use a sqlite database, JSON, YAML, ...?
- A lot less work for the end-user.
- Allows storing additional data, such as `RORG`, `FUNC`, `TYPE` for parsing in the future.
  - Would allow parsing of messages, without the need for user knowing the details.
  - This data could be further extended as we go, allowing "learning" of supported functionalities etc...
- Not sure if this is the job of this library, as it was originally meant as a pure library, with the end-user  creating the functionality. Now, with automatic teach-in responses (hopefully) this "border" is starting to fade...

_Number 3_
- Not sure we have a number 3?


**Generic notes**
- We should implement some kind of timing tests.
  - How long does it to parse, create, build a `Packet`?
  - How long does it to create a response packet to match a teach-in?
- These tests could be then replicated on low-end hardware like the Raspberry Pi, to see if we can stick to the time constraints set by the specification.
  - If not, optimise, optimise, optimise... Though this should be done anyways, I know at least a few places to look :stuck_out_tongue_closed_eyes:

**Comment by romor** [comment]
To give you an indication on timing: 
I use this library on the RPi B+ (thus, the old, slower model), which is also used for a couple of other tasks and enocean being more a side topic. To respond to an incoming request packet takes 500 ms (determined by the logging timestamps). To answer a learn request takes about the same time.

**Comment by alan-martinovic** [comment]
Is this issue closed? It seems to have spread a lot?
I am successfully communicating with an EnOcean device using 
https://github.com/kipe/enocean/pull/22/files#diff-4d1cac9a9a0b3cc9d71b8e7fef005dc4R46
as the transmitter id.

NOTE:
It seems that the EnOcean device I'm controlling always sends broadcast messages.
I've talked to EnOcean support and they said that is vendor specific behaviour.

**Comment by kipe** [comment]
You're correct, this issue has spread quite a bit. I'll create new issue(s) for more specific discussion tonight and close this one.

**Comment by kipe** [comment]
Not closing this yet, as it's still valid. But now on, lets move the other discussions to the corresponding issues.

This issue is still valid, as for determining the correct address for sending messages, we need to fetch the module id. This should be done automatically, when starting the communicator. After that, it should be used as the base of every message send. If required, the device could be sent a `Sender ID` derived from the `Base ID`, but `Destination ID` should always be set to the correct device address.
This does require testing on hardware, to make sure that setting the `Destination ID` doesn't mess up anything...

**Comment by alan-martinovic** [comment]
'module id' is the address of the EnOcean controller attached to a gateway?

**Comment by kipe** [comment]
Yes. EnOcean documentation refers it to as `Base ID`, if I remember correctly.

**Comment by romor** [comment]
> but Destination ID should always be set to the correct device address.

Not sure on this. EnOceanSerialProtocol3 describes the Destination ID as:

> Broadcast radio: FF FF FF FF
> ADT radio: Destination ID (= address)

All my devices (incl. the Thermokon message server communicating with the HVAC clients) use the broadcast address in their packets (at least this is what our enocean library tells me). 

I guess it does not matter what to use and I also don't know which approach woud be better.

**Comment by kipe** [comment]
Hmm, I think we should go with what the documentation says, even if it doesn't matter.
So unless the device supports ADT, let's use the broadcast address.

**Comment by kipe** [comment]
The fetching of the Base ID is now integrated to `Communicator`. Example is updated to correspond.

Still keeping this open though, so the last few comments aren't forgotten...
